File: local_examples/cmds_generic/predis/CmdsGenericTest.php
// EXAMPLE: cmds_generic
<?php
use PHPUnit\Framework\TestCase;
use Predis\Client as PredisClient;

class CmdsGenericTest
// REMOVE_START
extends TestCase
// REMOVE_END
{
    public function testCmdsGeneric() {
        $r = new PredisClient([
            'scheme'   => 'tcp',
            'host'     => '127.0.0.1',
            'port'     => 6379,
            'password' => '',
            'database' => 0,
        ]);
        // REMOVE_START
        $r->flushall();
        // REMOVE_END

        // STEP_START exists
        $existsResult1 = $r->set('key1', 'Hello');
        echo $existsResult1 . PHP_EOL; // >>> OK

        $existsResult2 = $r->exists('key1');
        echo $existsResult2 . PHP_EOL; // >>> 1

        $existsResult3 = $r->exists('nosuchkey');
        echo $existsResult3 . PHP_EOL; // >>> 0

        $existsResult4 = $r->set('key2', 'World');
        echo $existsResult4 . PHP_EOL; // >>> OK

        $existsResult5 = $r->exists('key1', 'key2', 'nosuchkey');
        echo $existsResult5 . PHP_EOL; // >>> 2
        // STEP_END
        // REMOVE_START
        $this->assertEquals('OK', $existsResult1);
        $this->assertEquals(1, $existsResult2);
        $this->assertEquals(0, $existsResult3);
        $this->assertEquals('OK', $existsResult4);
        $this->assertEquals(2, $existsResult5);
        $r->del('key1', 'key2');
        // REMOVE_END

        // STEP_START keys
        $keysResult1 = $r->mset(['firstname' => 'Jack', 'lastname' => 'Stuntman', 'age' => '35']);
        echo $keysResult1 . PHP_EOL; // >>> OK

        $keysResult2 = $r->keys('*name*');
        sort($keysResult2);
        echo implode(', ', $keysResult2) . PHP_EOL; // >>> firstname, lastname

        $keysResult3 = $r->keys('a??');
        echo implode(', ', $keysResult3) . PHP_EOL; // >>> age

        $keysResult4 = $r->keys('*');
        sort($keysResult4);
        echo implode(', ', $keysResult4) . PHP_EOL; // >>> age, firstname, lastname
        // STEP_END
        // REMOVE_START
        $this->assertEquals('OK', $keysResult1);
        $this->assertEquals(['firstname', 'lastname'], $keysResult2);
        $this->assertEquals(['age'], $keysResult3);
        $this->assertEquals(['age', 'firstname', 'lastname'], $keysResult4);
        $r->del('firstname', 'lastname', 'age');
        // REMOVE_END

        // STEP_START scan1
        $scan1Result1 = $r->sadd('myset', ['1', '2', '3', 'foo', 'foobar', 'feelsgood']);
        echo $scan1Result1 . PHP_EOL;                       // >>> 6

        [$scan1Cursor, $scan1Members] = $r->sscan('myset', 0, ['MATCH' => 'f*']);
        sort($scan1Members);
        echo implode(', ', $scan1Members) . PHP_EOL;         // >>> feelsgood, foo, foobar
        // STEP_END

        // REMOVE_START
        $this->assertEquals(6, $scan1Result1);
        $this->assertEquals(['feelsgood', 'foo', 'foobar'], $scan1Members);
        $r->del('myset');
        // REMOVE_END

        // STEP_START scan2
        // REMOVE_START
        for ($i = 1; $i <= 1000; $i++) {
            $r->set("key:$i", $i);
        }
        // REMOVE_END

        // MATCH is applied after elements are fetched, so with the default COUNT most
        // iterations return few keys or none at all.
        $scan2Cursor = 0;

        for ($i = 0; $i < 4; $i++) {
            [$scan2Cursor, $scan2Keys] = $r->scan($scan2Cursor, ['MATCH' => '*11*']);
            echo count($scan2Keys) . PHP_EOL;
        }

        // A larger COUNT forces more scanning in a single iteration, so the remaining
        // matches arrive together. This continues from the cursor reached above.
        [$scan2Cursor, $scan2Keys] = $r->scan($scan2Cursor, ['MATCH' => '*11*', 'COUNT' => 1000]);
        echo count($scan2Keys) . PHP_EOL;                    // >>> 18
        // STEP_END

        // REMOVE_START
        $this->assertEquals(18, count($scan2Keys));
        $r->flushdb();
        // REMOVE_END

        // STEP_START scan4
        $scan4Result1 = $r->hset('myhash', 'a', 1, 'b', 2);
        echo $scan4Result1 . PHP_EOL;                        // >>> 2

        [$scan4Cursor, $scan4Pairs] = $r->hscan('myhash', 0);
        echo json_encode($scan4Pairs) . PHP_EOL;             // >>> {"a":"1","b":"2"}

        // Redis does not promise a field order, so sort before comparing.
        [$scan4Cursor, $scan4Fields] = $r->hscan('myhash', 0, ['NOVALUES' => true]);
        sort($scan4Fields);
        echo implode(', ', $scan4Fields) . PHP_EOL;          // >>> a, b
        // STEP_END

        // REMOVE_START
        $this->assertEquals(2, $scan4Result1);
        $this->assertEquals(['a' => '1', 'b' => '2'], $scan4Pairs);
        $this->assertEquals(['a', 'b'], $scan4Fields);
        $r->del('myhash');
        // REMOVE_END

        // STEP_START del
        echo $r->set('key1', 'Hello') . PHP_EOL;             // >>> OK
        echo $r->set('key2', 'World') . PHP_EOL;             // >>> OK

        $delResult = $r->del('key1', 'key2', 'key3');
        echo $delResult . PHP_EOL;                           // >>> 2
        // STEP_END

        // REMOVE_START
        $this->assertEquals(2, $delResult);
        // REMOVE_END

        // STEP_START expire
        echo $r->set('mykey', 'Hello') . PHP_EOL;            // >>> OK

        echo $r->expire('mykey', 10) . PHP_EOL;              // >>> 1
        echo $r->ttl('mykey') . PHP_EOL;                     // >>> 10

        // Overwriting a key with SET clears its expiry.
        echo $r->set('mykey', 'Hello World') . PHP_EOL;      // >>> OK
        echo $r->ttl('mykey') . PHP_EOL;                     // >>> -1

        // XX only sets the expiry when one already exists, so this is a no-op.
        echo $r->expire('mykey', 10, 'XX') . PHP_EOL;        // >>> 0
        echo $r->ttl('mykey') . PHP_EOL;                     // >>> -1

        // NX only sets the expiry when there is none, so this one applies.
        echo $r->expire('mykey', 10, 'NX') . PHP_EOL;        // >>> 1
        $expireTtl = $r->ttl('mykey');
        echo $expireTtl . PHP_EOL;                           // >>> 10
        // STEP_END

        // REMOVE_START
        $this->assertEquals(10, $expireTtl);
        $r->del('mykey');
        // REMOVE_END

        // STEP_START ttl
        echo $r->set('mykey', 'Hello') . PHP_EOL;            // >>> OK
        echo $r->expire('mykey', 10) . PHP_EOL;              // >>> 1

        $ttlResult = $r->ttl('mykey');
        echo $ttlResult . PHP_EOL;                           // >>> 10
        // STEP_END

        // REMOVE_START
        $this->assertEquals(10, $ttlResult);
        $r->del('mykey');
        // REMOVE_END

    }
}

File: local_examples/cmds_generic/go-redis/cmds_generic_test.go
// EXAMPLE: cmds_generic
// HIDE_START
package example_commands_test

import (
	"context"
	"fmt"
	"math"
	"sort"
	"time"

	"github.com/redis/go-redis/v9"
)

// HIDE_END

func ExampleClient_del_cmd() {
	ctx := context.Background()

	rdb := redis.NewClient(&redis.Options{
		Addr:     "localhost:6379",
		Password: "", // no password docs
		DB:       0,  // use default DB
	})

	// REMOVE_START
	// make sure we are working with fresh database
	rdb.FlushDB(ctx)
	rdb.Del(ctx, "key1", "key2", "key3")
	// REMOVE_END

	// STEP_START del
	delResult1, err := rdb.Set(ctx, "key1", "Hello", 0).Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(delResult1) // >>> OK

	delResult2, err := rdb.Set(ctx, "key2", "World", 0).Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(delResult2) // >>> OK

	delResult3, err := rdb.Del(ctx, "key1", "key2", "key3").Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(delResult3) // >>> 2
	// STEP_END

	// Output:
	// OK
	// OK
	// 2
}

func ExampleClient_exists_cmd() {
	ctx := context.Background()

	rdb := redis.NewClient(&redis.Options{
		Addr:     "localhost:6379",
		Password: "", // no password docs
		DB:       0,  // use default DB
	})

	// REMOVE_START
	// make sure we are working with fresh database
	rdb.FlushDB(ctx)
	rdb.Del(ctx, "key1", "key2", "nosuchkey")
	// REMOVE_END

	// STEP_START exists
	existsResult1, err := rdb.Set(ctx, "key1", "Hello", 0).Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(existsResult1) // >>> OK

	existsResult2, err := rdb.Exists(ctx, "key1").Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(existsResult2) // >>> 1

	existsResult3, err := rdb.Exists(ctx, "nosuchkey").Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(existsResult3) // >>> 0

	existsResult4, err := rdb.Set(ctx, "key2", "World", 0).Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(existsResult4) // >>> OK

	existsResult5, err := rdb.Exists(ctx, "key1", "key2", "nosuchkey").Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(existsResult5) // >>> 2
	// STEP_END

	// Output:
	// OK
	// 1
	// 0
	// OK
	// 2
}

func ExampleClient_expire_cmd() {
	ctx := context.Background()

	rdb := redis.NewClient(&redis.Options{
		Addr:     "localhost:6379",
		Password: "", // no password docs
		DB:       0,  // use default DB
	})

	// REMOVE_START
	// start with fresh database
	rdb.FlushDB(ctx)
	rdb.Del(ctx, "mykey")
	// REMOVE_END

	// STEP_START expire
	expireResult1, err := rdb.Set(ctx, "mykey", "Hello", 0).Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(expireResult1) // >>> OK

	expireResult2, err := rdb.Expire(ctx, "mykey", 10*time.Second).Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(expireResult2) // >>> true

	expireResult3, err := rdb.TTL(ctx, "mykey").Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(math.Round(expireResult3.Seconds())) // >>> 10

	expireResult4, err := rdb.Set(ctx, "mykey", "Hello World", 0).Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(expireResult4) // >>> OK

	expireResult5, err := rdb.TTL(ctx, "mykey").Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(expireResult5) // >>> -1ns

	expireResult6, err := rdb.ExpireXX(ctx, "mykey", 10*time.Second).Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(expireResult6) // >>> false

	expireResult7, err := rdb.TTL(ctx, "mykey").Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(expireResult7) // >>> -1ns

	expireResult8, err := rdb.ExpireNX(ctx, "mykey", 10*time.Second).Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(expireResult8) // >>> true

	expireResult9, err := rdb.TTL(ctx, "mykey").Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(math.Round(expireResult9.Seconds())) // >>> 10
	// STEP_END

	// Output:
	// OK
	// true
	// 10
	// OK
	// -1ns
	// false
	// -1ns
	// true
	// 10
}

func ExampleClient_keys_cmd() {
	ctx := context.Background()

	rdb := redis.NewClient(&redis.Options{
		Addr:     "localhost:6379",
		Password: "", // no password docs
		DB:       0,  // use default DB
	})

	// REMOVE_START
	rdb.FlushDB(ctx)
	// REMOVE_END

	// STEP_START keys
	keysResult1, err := rdb.MSet(ctx, "firstname", "Jack", "lastname", "Stuntman", "age", "35").Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(keysResult1) // >>> OK

	keysResult2, err := rdb.Keys(ctx, "*name*").Result()

	if err != nil {
		panic(err)
	}

	sort.Strings(keysResult2)
	fmt.Println(keysResult2) // >>> [firstname lastname]

	keysResult3, err := rdb.Keys(ctx, "a??").Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(keysResult3) // >>> [age]

	keysResult4, err := rdb.Keys(ctx, "*").Result()

	if err != nil {
		panic(err)
	}

	sort.Strings(keysResult4)
	fmt.Println(keysResult4) // >>> [age firstname lastname]
	// STEP_END

	// Output:
	// OK
	// [firstname lastname]
	// [age]
	// [age firstname lastname]
}

func ExampleClient_ttl_cmd() {
	ctx := context.Background()

	rdb := redis.NewClient(&redis.Options{
		Addr:     "localhost:6379",
		Password: "", // no password docs
		DB:       0,  // use default DB
	})

	// REMOVE_START
	// start with fresh database
	rdb.FlushDB(ctx)
	rdb.Del(ctx, "mykey")
	// REMOVE_END

	// STEP_START ttl
	ttlResult1, err := rdb.Set(ctx, "mykey", "Hello", 10*time.Second).Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(ttlResult1) // >>> OK

	ttlResult2, err := rdb.TTL(ctx, "mykey").Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(math.Round(ttlResult2.Seconds())) // >>> 10
	// STEP_END

	// Output:
	// OK
	// 10
}

func ExampleClient_scan1_cmd() {
	ctx := context.Background()

	rdb := redis.NewClient(&redis.Options{
		Addr:     "localhost:6379",
		Password: "", // no password docs
		DB:       0,  // use default DB
	})

	// REMOVE_START
	rdb.FlushDB(ctx)
	// REMOVE_END

	// STEP_START scan1
	scan1Result1, err := rdb.SAdd(ctx, "myset", "1", "2", "3", "foo", "foobar", "feelsgood").Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(scan1Result1) // >>> 6

	scan1Result2, _, err := rdb.SScan(ctx, "myset", 0, "f*", 0).Result()

	if err != nil {
		panic(err)
	}

	sort.Strings(scan1Result2)
	fmt.Println(scan1Result2) // >>> [feelsgood foo foobar]
	// STEP_END

	// Output:
	// 6
	// [feelsgood foo foobar]
}

func ExampleClient_scan2_cmd() {
	ctx := context.Background()

	rdb := redis.NewClient(&redis.Options{
		Addr:     "localhost:6379",
		Password: "", // no password docs
		DB:       0,  // use default DB
	})

	// REMOVE_START
	rdb.FlushDB(ctx)

	for i := 1; i <= 1000; i++ {
		rdb.Set(ctx, fmt.Sprintf("key:%d", i), i, 0)
	}
	// REMOVE_END

	// STEP_START scan2
	// MATCH is applied after elements are fetched, so with the default COUNT most
	// iterations return few keys or none at all.
	var scan2Cursor uint64
	var scan2Keys []string
	var err error
	scan2Total := 0

	for i := 0; i < 4; i++ {
		scan2Keys, scan2Cursor, err = rdb.Scan(ctx, scan2Cursor, "*11*", 0).Result()

		if err != nil {
			panic(err)
		}
		scan2Total += len(scan2Keys)
	}

	// A larger COUNT forces more scanning in a single iteration, so the remaining
	// matches arrive together. This continues from the cursor reached above.
	scan2Keys, _, err = rdb.Scan(ctx, scan2Cursor, "*11*", 1000).Result()

	if err != nil {
		panic(err)
	}
	scan2Total += len(scan2Keys)

	// The per-call split isn't guaranteed, but the cumulative total is.
	fmt.Println(scan2Total) // >>> 19
	// STEP_END

	// REMOVE_START
	rdb.FlushDB(ctx)
	// REMOVE_END

	// Output:
	// 19
}

func ExampleClient_scan3_cmd() {
	ctx := context.Background()

	rdb := redis.NewClient(&redis.Options{
		Addr:     "localhost:6379",
		Password: "", // no password docs
		DB:       0,  // use default DB
	})

	// REMOVE_START
	rdb.FlushDB(ctx)
	// REMOVE_END

	// STEP_START scan3
	scan3Result1, err := rdb.GeoAdd(ctx, "geokey", &redis.GeoLocation{
		Longitude: 0, Latitude: 0, Name: "value",
	}).Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(scan3Result1) // >>> 1

	scan3Result2, err := rdb.ZAdd(ctx, "zkey", redis.Z{Score: 1000, Member: "value"}).Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(scan3Result2) // >>> 1

	scan3Result3, err := rdb.Type(ctx, "geokey").Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(scan3Result3) // >>> zset

	scan3Result4, err := rdb.Type(ctx, "zkey").Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(scan3Result4) // >>> zset

	// A single call isn't guaranteed to find every match, so loop until the cursor
	// returns to 0, accumulating matches from every call.
	var scan3Cursor uint64
	var scan3Keys []string
	var scan3Batch []string

	for {
		scan3Batch, scan3Cursor, err = rdb.ScanType(ctx, scan3Cursor, "", 0, "zset").Result()

		if err != nil {
			panic(err)
		}
		scan3Keys = append(scan3Keys, scan3Batch...)

		if scan3Cursor == 0 {
			break
		}
	}

	sort.Strings(scan3Keys)
	fmt.Println(scan3Keys) // >>> [geokey zkey]
	// STEP_END

	// Output:
	// 1
	// 1
	// zset
	// zset
	// [geokey zkey]
}

func ExampleClient_scan4_cmd() {
	ctx := context.Background()

	rdb := redis.NewClient(&redis.Options{
		Addr:     "localhost:6379",
		Password: "", // no password docs
		DB:       0,  // use default DB
	})

	// REMOVE_START
	rdb.FlushDB(ctx)
	// REMOVE_END

	// STEP_START scan4
	scan4Result1, err := rdb.HSet(ctx, "myhash", "a", 1, "b", 2).Result()

	if err != nil {
		panic(err)
	}

	fmt.Println(scan4Result1) // >>> 2

	scan4Result2, _, err := rdb.HScan(ctx, "myhash", 0, "", 0).Result()

	if err != nil {
		panic(err)
	}

	// HSCAN returns field and value interleaved. Redis does not promise an order, so
	// collect the pairs into a map: fmt prints map keys sorted, whatever order they arrived in.
	scan4Fields := map[string]string{}

	for i := 0; i < len(scan4Result2); i += 2 {
		scan4Fields[scan4Result2[i]] = scan4Result2[i+1]
	}

	fmt.Println(scan4Fields) // >>> map[a:1 b:2]

	scan4Result3, _, err := rdb.HScanNoValues(ctx, "myhash", 0, "", 0).Result()

	if err != nil {
		panic(err)
	}

	sort.Strings(scan4Result3)
	fmt.Println(scan4Result3) // >>> [a b]
	// STEP_END

	// Output:
	// 2
	// map[a:1 b:2]
	// [a b]
}
